ext/drreg/drreg.c

```diff
@@ -938,47 +1627,84 @@ drreg_restore_app_value(void *drcontext, instrlist_t *ilist, instr_t *where,
+                pt->reg[GPR_IDX(app_reg)].native = true;
         }
-        instrlist_set_auto_predicate(ilist, pred);
-        return DRREG_SUCCESS;
     }
+#ifdef X86
+    else if (reg_is_vector_simd(app_reg)) {
+        if (!reg_is_vector_simd(dst_reg))
+            return DRREG_ERROR_INVALID_PARAMETER;
 
-    /* we may have lost the app value for a dead reg */
-    if (!pt->reg[GPR_IDX(app_reg)].ever_spilled) {
-        LOG(drcontext, DR_LOG_ALL, 3, "%s @%d." PFX ": reg %s never spilled\n",
-            __FUNCTION__, pt->live_idx, get_where_app_pc(where),
-            get_register_name(app_reg));
-        instrlist_set_auto_predicate(ilist, pred);
-        return DRREG_ERROR_NO_APP_VALUE;
-    }
-    /* restore app value back to app_reg */
-    if (pt->reg[GPR_IDX(app_reg)].xchg != DR_REG_NULL) {
-        /* XXX i#511: NYI */
-        instrlist_set_auto_predicate(ilist, pred);
-        return DRREG_ERROR_FEATURE_NOT_AVAILABLE;
-    }
-    LOG(drcontext, DR_LOG_ALL, 3, "%s @%d." PFX ": getting app value for %s\n",
-        __FUNCTION__, pt->live_idx, get_where_app_pc(where), get_register_name(app_reg));
-    /* XXX i#511: if we add .xchg support for GPR's we'll need to check them all here. */
-    if (pt->aflags.xchg == app_reg) {
-        /* Bail on keeping the flags in the reg. */
-        drreg_move_aflags_from_reg(drcontext, ilist, where, pt, stateful);
-    } else {
-        restore_reg(drcontext, pt, app_reg, pt->reg[GPR_IDX(app_reg)].slot, ilist, where,
-                    stateful && !pt->reg[GPR_IDX(app_reg)].in_use);
-        if (stateful && !pt->reg[GPR_IDX(app_reg)].in_use)
-            pt->reg[GPR_IDX(app_reg)].native = true;
+        /* check if app_reg is an unspilled reg */
+        if (pt->simd_reg[SIMD_IDX(app_reg)].native) {
+            if (dst_reg != app_reg) {
+                PRE(ilist, where,
+                    INSTR_CREATE_movdqa(drcontext, opnd_create_reg(dst_reg),
+                                        opnd_create_reg(app_reg)));
+            }
+            instrlist_set_auto_predicate(ilist, pred);
+            return DRREG_SUCCESS;
+        }
+        /* we may have lost the app value for a dead reg */
+        if (!pt->simd_reg[SIMD_IDX(app_reg)].ever_spilled) {
+            instrlist_set_auto_predicate(ilist, pred);
+            return DRREG_ERROR_NO_APP_VALUE;
+        }
+        /* restore app value back to app_reg */
+        if (pt->simd_reg[SIMD_IDX(app_reg)].xchg != DR_REG_NULL) {
+            /* XXX i#511: NYI */
+            instrlist_set_auto_predicate(ilist, pred);
+            return DRREG_ERROR_FEATURE_NOT_AVAILABLE;
+        }
+        restore_reg_indirectly(drcontext, pt, app_reg,
+                               pt->simd_reg[SIMD_IDX(app_reg)].slot, ilist, where,
+                               stateful && !pt->simd_reg[SIMD_IDX(app_reg)].in_use);
+        if (stateful && !pt->simd_reg[SIMD_IDX(app_reg)].in_use)
+            pt->simd_reg[SIMD_IDX(app_reg)].native = true;
```

Comment:
Again, I'm seeing big blocks of code duplicated with just different arrays or slight differences.  Was sharing already considered?  If so and it was rejected, please add comments on why it's not shared for future readers who wonder at these multiple sets of duplicated code blocks.

Reply:
Same comments as above.